design/design-doc-agent.md

```diff
@@ -0,0 +1,586 @@
+# Crossplane Agent for Consumption
+
+* Owner: Muvaffak Onuş (@muvaf)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+Crossplane allows users to provision and manage cloud services from your
+Kubernetes cluster. It has managed resources that map to the services in the
+provider 1-to-1 as the lowest level resource. Then users can build & publish
+their own APIs that are abstractions over these managed resources. The actual
+connection and consumption of these resources by applications are handled with
+namespaced types called requirements whose CRDs are created via
+`InfrastructurePublication`s and have `*Requirement` suffix in their kind.
+
+The consumption model is such that applications should create requirements to
+request certain services and supply `Secret` name in the requirements which will
+be used by Crossplane to populate the necessary credentials for application to
+consume. As a simple example, an application bundle would have a
+`MySQLInstanceRequirement` custom resource, a `Pod` and they would share the
+same name for the secret so that one fills that `Secret` with credentials and
+the other one mounts it for the containers to consume.
+
+> For brevity, application will be assumed to have only one `Pod`.
+
+This consumption model works well in cases where you deploy your application in
+a cluster where Crossplane runs, i.e. local mode. However, there are many valid
+use cases where this is not really feasible. Here is a subset of those use
+cases:
+
+* Private Networking.
+  * You may want to deploy different applications into different VPCs but manage
+    all of your infrastructure from one place. This isn't possible since you are
+    deploying all applications into the same cluster to have them use Crossplane
+    and being in the same cluster necessitates usage of the same VPC.
+* Cluster Configuration.
+  * Because you have to run applications in the same central cluster with
+    others, you'll have to share the same Kubernetes resources like nodes and
+    your needs in terms of instance types could differ greatly depending on your
+    workloads, like some need GPU-powered machines and others memory-heavy ones.
+* Security.
+    * All applications are subject to the same user management domain, i.e. same
+      api-server. This could be managed to be safe, but it's not physically
+      impossible to have a `ServiceAccount` in another namespace to have access
+      to resources in your namespace. So, you wouldn't really trust to have
+      production in one namespace and dev in the other.
```

Comment:
Actually I was meaning to refer to the native Kubernetes resources here instead of Crossplane ones because the comparison is about local and remote scenarios, and that comparison doesn't really change separation of namespace-scoped and cluster-scoped objects being more secure or not. It's more about multi-namespace usage of Kubernetes as workload scheduler. Obviously, it works well for a lot of cases but still the possibility of escape does exist, like it's not an isolation level that GCP Project provides or isolation level that you'd have in VMs that run on the same machine. That's why people use different projects for prod workloads for example.